0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+    }
+  },
+  "additionalProperties": false,
+  "required": [
+    "metadata"
+  ]
+}
+```
+
+If there is no `id` JSON property the request payload than server will process the request but will not publish the status response back to client.
+The server MUST publish the [status response](#full-metadata-update-response) with the `id` property in the payload and processing status fields.
+
+Examples:
+- update metadata with message identifier
+
+```json
+{
+  "id":42,
+  "metadata":{
+    "deviceModel":"example model",
+    "name":"Sensor 1"
+  }
+}
+```
+- update metadata without message identifier
+
+```json
+{
+  "metadata":{
+    "deviceModel":"example model",
+    "name":"Sensor 1"
+  }
+}
+```
+
+#### Full metadata update response
+
+The server MUST respond to the metadata get request by publishing the response message according to the [request/response design pattern defined in 1/KP](/0001-kaa-protocol/README.md#requestresponse-pattern).
+The extension-specific resource path part format is:
+
+  `<endpoint_token>/update/status`
+
+The response payload MUST be a JSON-encoded object with the following [JSON schema](http://json-schema.org/) ([file](./schemas/update-metadata-response.schema.json)):
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "10/EPMDX update metadata response",
+  "type": "object",
+  "properties": {
+    "id": {
+      "type": "integer",
+      "description": "The message identifier used to match server response to the request"
+    },
+    "statusCode": {
+      "type": "integer",
+      "minimum": 1,
+      "description": "Status code based on HTTP status codes"
+    },
+    "reasonPhrase": {
+      "type": "string",
+      "minLength": 1,
+      "description": "A human-readable string explaining the cause of an error (if any). `OK` if request was successful"
+    }
+  },
+  "required": [
+    "statusCode",
+    "reasonPhrase"
+  ],
+  "additionalProperties": false
+}
+```
+
+  where
+  - `id` MUST match that in the request
+
+Examples:
+
+- Endpoint metadata successfully updated
+
+```json
+{
+  "id":42,
+  "statusCode":200,
+  "reasonPhrase":"OK"
+}
+```
+- Endpoint metadata successfully created
+
+```json
+{
+  "id":42,
+  "statusCode":201,
+  "reasonPhrase":"Created"
+}
+```
+### Partial metadata update
+
+Create or update endpoint metadata key-value pairs.
+In case of update the provided metadata will override all existing metadata key-value pairs.
+In comparison with [full metadata update operation](#full-metadata-update) this operation will not override all metadata key-value pairs.
+It will only create or update the metadata keys specified in the request.
+
+#### Partial metadata update request
+
+Client MUST send requests with the following extension-specific [resource path](/0001-kaa-protocol/README.md#resource-path-format) part in order to receive the metadata key-value pairs:
+
+  `<endpoint_token>/update/keys`
```

Comment:
The `replace` path may be confusing because the replace operation is supposed to be executed on already existing metadata key-value pair. But in partial update we support `create or update` of key-value pair. 
I'm thinking on other thing: do we have a use-case when endpoint should perform full update instead of partial?